### PR TITLE
Fix reflection on bitfield.

### PR DESCRIPTION
**Pull request by python3kgae**
This is for case where bitfield member lost when reflection.

Save bitfield information in DxilFieldAnnotation.
Then use the information in CShaderReflectionType::Initialize.

### CODE REVIEW

File: include/dxc/DXIL/DxilTypeSystem.h [hunk 1]
```diff
@@ -96,6 +104,8 @@ class DxilFieldAnnotation {
   InterpolationMode m_InterpMode;
   std::string m_FieldName;
   bool m_bCBufferVarUsed; // true if this field represents a top level variable in CB structure, and it is used.
+  std::vector<DxilFieldAnnotation> m_BitFields;
+  unsigned m_BitFieldWidth; // For bit field. 0 means not bitfiled.
```
Comment (tex3d) [inline include/dxc/DXIL/DxilTypeSystem.h:108]: Comment typo: `bitfield` rather than `bitfiled`

Comment (python3kgae) [reply include/dxc/DXIL/DxilTypeSystem.h]: Fixed.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl [hunk 2]
```diff
@@ -0,0 +1,77 @@
+// RUN: %dxc -E main -T ps_6_0 -HV 2021 -Vd -validator-version 0.0 %s | %D3DReflect %s | FileCheck %s
+
+// Make sure bitfiled info is saved.
+// TODO - make better dump. Now, Rows = 0 means bitfield. Columns is bit width. Elements is bit offset.
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:4]: Comment out of date

File: tools/clang/unittests/HLSLTestLib/D3DReflectionStrings.cpp [hunk 3]
```diff
@@ -198,7 +198,18 @@ LPCSTR ToString(D3D_TESSELLATOR_DOMAIN TessellatorDomain) {
   default: return nullptr;
   }
 }
+
+// FIXME: remove this once D3D_SVC_BIT_FIELD added into
+// D3D_SHADER_VARIABLE_CLASS.
+const D3D_SHADER_VARIABLE_CLASS D3D_SVC_BIT_FIELD =
```
Comment (tex3d) [inline tools/clang/unittests/HLSLTestLib/D3DReflectionStrings.cpp:204]: See how values are added for 16 and 64 bit types under `ADD_16_64_BIT_TYPES`.  Maybe we should follow the same pattern.  It allows for compilation against the header after you add it as well as before.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl (hunk 2 above)
Comment (tex3d) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:4]: To clarify:
- `Rows = 0 means bitfield` shouldn't be the indicator, since it uses `D3D_SVC_BIT_FIELD` now.  However, I still see that Rows are 0 for bitfields.  Should they be?  I'm not sure it matters, but it's not necessary anymore.
- `Elements is bit offset` - this is `Offset` now.

File: lib/HLSL/DxilContainerReflection.cpp [hunk 4]
```diff
@@ -102,6 +103,14 @@ enum class PublicAPI { D3D12 = 0, D3D11_47 = 1, D3D11_43 = 2 };
 #define D3D_SVT_UINT64  ((D3D_SHADER_VARIABLE_TYPE)62)
 #endif // ADD_16_64_BIT_TYPES
 
+#ifdef ADD_SVC_BIT_FIELD
+// Disable warning about value not being valid in enum
+#pragma warning(disable : 4063)
+// FIXME: remove the define once D3D_SVC_BIT_FIELD added into
+// D3D_SHADER_VARIABLE_CLASS.
+#define D3D_SVC_BIT_FIELD (D3D_SVC_INTERFACE_POINTER + 1)
```
Comment (tex3d) [inline lib/HLSL/DxilContainerReflection.cpp:111]: I believe this should have a cast inside this expression.
```cpp
#define D3D_SVC_BIT_FIELD ((D3D_SHADER_VARIABLE_CLASS)(D3D_SVC_INTERFACE_POINTER + 1))
```

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl (hunk 2 above)
Comment (pow2clk) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:4]: Xiang, are you going to address this comment?

Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl]: Fixed. The outdated comment is removed. Rows will be 1 now.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+// CHECK-NEXT:                 }
+// CHECK-NEXT:             }
+// CHECK-NEXT: CBuffer: B
+
+struct BF {
+   int i0 : 8;
+   int i1 : 8;
+   int i2 : 8;
+   int i3 : 8;
+};
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:69]: I think we need more test cases. There are places where I was trying to think how it might mess up other elements of the struct. Can we get tests where the bitfield doesn't fully fill the integer? One where a zero width ends the run? one with a non-bitfield member at the start? Maybe one with a single member with a bitfield width?

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 6]
```diff
@@ -985,6 +1001,12 @@ unsigned CGMSHLSLRuntime::ConstructStructAnnotation(DxilStructAnnotation *annota
           ConstructFieldAttributedAnnotation(fieldAnnotation, fieldTy,
                                              bDefaultRowMajor);
           fieldAnnotation.SetCBufferOffset(CBufferOffset);
+          if (StartBitOffset > 0) {
+            DxilFieldAnnotation &prevFieldAnnotation =
+                annotation->GetFieldAnnotation(fieldIdx - 1);
```
Comment (pow2clk) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp:1006]: Can `fieldIdx` be zero? What if the struct is just one member with a bitfield width?

**Review by pow2clk — COMMENTED** [review]
Most of my comments are nits, but I have some concerns about subtracting one from the index. I would like to better understand the idea here. The first bitfielded member will have an array containing all the others that make up the full subdivided value? is that right?

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp (hunk 6 above)
Comment (python3kgae) [reply tools/clang/lib/CodeGen/CGHLSLMS.cpp:1006]: If fieldIdx is zero then, StartBitOffset will be zero too.

**Comment by python3kgae** [comment]
> Most of my comments are nits, but I have some concerns about subtracting one from the index. I would like to better understand the idea here. The first bitfielded member will have an array containing all the others that make up the full subdivided value? is that right?

There will be a fake field created for the whole integer. Then all fields are saved as members of that fake field.

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 7]
```diff
@@ -958,15 +958,31 @@ unsigned CGMSHLSLRuntime::ConstructStructAnnotation(DxilStructAnnotation *annota
...
+        DxilFieldAnnotation bitfieldAnnotation;
+
+        bitfieldAnnotation.SetBitFieldWidth(Field->getBitWidthValue(Context));
+        const BuiltinType *BTy = Field->getType()->getAs<BuiltinType>();
+        CompType::Kind kind =
+            BuiltinTyToCompTy(BTy, /*bSNorm*/ false, /*bUNorm*/ false);
+        bitfieldAnnotation.SetCompType(kind);
+        bitfieldAnnotation.SetFieldName(Field->getName());
+        bitfieldAnnotation.SetCBufferOffset((unsigned)(BitOffset - StartBitOffset));
+        // TODO: more information for bitfield.
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply tools/clang/lib/CodeGen/CGHLSLMS.cpp:983]: Do we plan to support semantic on bit field?
It will be strange to have things like uint bf : SemanticA : 8 :)

File: lib/HLSL/DxilContainerReflection.cpp [hunk 8]
```diff
@@ -1169,15 +1185,47 @@ HRESULT CShaderReflectionType::Initialize(
...
+          unsigned bitOffset = 0;
+          CShaderReflectionType *bitFieldReflectionType = nullptr;
+          for (auto &bitfieldAnnotation : fieldAnnotation.GetBitFields()) {
+            bitFieldReflectionType = new CShaderReflectionType();
+            allTypes.push_back(
+                std::unique_ptr<CShaderReflectionType>(bitFieldReflectionType));
+
+            bitFieldReflectionType->Initialize(M, fieldType, fieldAnnotation,
+                                            elementOffset, allTypes, isCBuffer);
+            // FIXME: remove the cast once D3D_SVC_BIT_FIELD added into D3D_SHADER_VARIABLE_CLASS.
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply lib/HLSL/DxilContainerReflection.cpp:1214]: Created https://github.com/microsoft/DirectXShaderCompiler/issues/5169 to track this issue

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl (hunk 5 above)
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:69]: Added.

**Review by pow2clk — APPROVED** [review]
Just one comment nit. Looks good!

File: lib/DXIL/DxilMetadataHelper.cpp [hunk 9]
```diff
@@ -1214,6 +1214,20 @@ Metadata *DxilMDHelper::EmitDxilFieldAnnotation(const DxilFieldAnnotation &FA) {
...
+    for (const DxilFieldAnnotation &BitField : BitFields) {
+      MDBitFieldVals.emplace_back(EmitDxilFieldAnnotation(BitField));
+    }
+    auto *BitFieldsMD = MDNode::get(m_Ctx, MDBitFieldVals);
+    MDVals.emplace_back(Uint32ToConstMD(kDxilFieldAnnotationBitFieldsTag));
+    MDVals.emplace_back(BitFieldsMD);
+  }
+  if (FA.HasBitFieldWidth()) {
+    MDVals.emplace_back(Uint32ToConstMD(kDxilFieldAnnotationBitFieldWidthTag));
+    MDVals.emplace_back(Uint32ToConstMD((unsigned)FA.GetBitFieldWidth()));
```
Comment (bob80905) [inline lib/DXIL/DxilMetadataHelper.cpp:1229]: why is the (unsigned) cast necessary? 
Doesn't GetBitFieldWidth() already return an unsigned type?

**Review by hekota — APPROVED** [review]
LGTM!

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl (hunk 5 above)
Comment (tex3d) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:69]: Oh, I think we need a test that demonstrates how bitfields spill over into additional primitive elements, and should have additional top-level type with sub-fields.  Also, ensuring that an unnamed padding field is handled properly, and the `:0` forces the next bit field to start at the next primitive.

For instance:
```hlsl
struct BF {
   int i0 : 8;
   int i1 : 8;
   int i2 : 8;
   int i3 : 8;
   int i4 : 8; // spill over, start of second primitive
   int : 4; // what about blank field?  do we support this?
   int i5 : 5;
   int : 0; // force next bit field to start at next primitive
   int i6 : 6; // start of third primitive
};
```

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:1]: More descriptive names would help.  For this, I think something like `bitfield-in-structured-buffer.hlsl` might be good.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection2.hlsl
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection2.hlsl:1]: Name could be better, maybe `bitfield-combine-to-unsigned.hlsl`?

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection3.hlsl
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection3.hlsl:1]: What does this test add?

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection4.hlsl
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection4.hlsl:1]: I don't understand what this test adds.  Also, the FIXME comment doesn't seem relevant to bitfield testing, is it?

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl [hunk 10]
```diff
@@ -0,0 +1,76 @@
+// RUN: %dxc -E main -T ps_6_0 -HV 2021 -Vd -validator-version 0.0 %s | %D3DReflect %s | FileCheck %s
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:1]: I think we might be able to switch to using `-select-validator internal` instead of `-Vd -validator-version 0.0`

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tex3d) [reply tools/clang/lib/CodeGen/CGHLSLMS.cpp]: I do not think we support (or should support) semantics on bitfields.  If you want bitfield in signature element (if we support that?), then you would need to wrap the bitfields in a struct and put the semantic on the field or parameter using that struct type.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl (hunk 5 above)
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:69]: Done.

File: lib/DXIL/DxilMetadataHelper.cpp (hunk 9 above)
Comment (python3kgae) [reply lib/DXIL/DxilMetadataHelper.cpp]: Good catch. Removed.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl (hunk 10 above)
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection.hlsl:1]: unknown argument :(

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection4.hlsl
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection4.hlsl:1]: Removed.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection3.hlsl
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_reflection3.hlsl:1]: test bit field as single member.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+// CHECK-NEXT:                       Class: D3D_SVC_BIT_FIELD
+// CHECK-NEXT:                       Type: D3D_SVT_INT
+// CHECK-NEXT:                       Elements: 0
+// CHECK-NEXT:                       Rows: 1
+// CHECK-NEXT:                       Columns: 8
+// CHECK-NEXT:                       Members: 0
+// CHECK-NEXT:                       Offset: 0
+// CHECK-NEXT:                   ID3D12ShaderReflectionType:
+// CHECK-NEXT:                     D3D12_SHADER_TYPE_DESC: Name: int
+// CHECK-NEXT:                       Class: D3D_SVC_BIT_FIELD
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl:83]: I'm torn.  Should padding be reflected here?  I don't suppose it hurts, but we also have offsets for all the named fields, so we don't technically need it, right?  When we add field names, I guess the name of this field would be nullptr, or "\0"?

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+// CHECK-NEXT:             }
+// CHECK-NEXT: CBuffer: B
+
+struct BF {
+   int i0 : 8;
+   int i1 : 8;
+   int i2 : 8;
+   int i3 : 8;
+   int i4 : 8; // spill over, start of second primitive
+   int : 4; // what about blank field?  do we support this?
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl:129]: Great, now we should drop the questions in the comment from my example, since they are answered ;)

Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl:129]: :). I'll remove it.

File: tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl (hunk 11 above)
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/d3dreflect/bitfield_spill-over.hlsl:83]: Add padding filed seems redundant like you mentioned.
For the name, I'm not sure. Both nullptr and "\0" feels OK.
